Rules/UseShouldProcessForStateChangingFunctions.cs

```diff
@@ -0,0 +1,126 @@
+using System.ComponentModel.Composition;
+using System.Globalization;
+
+namespace Microsoft.Windows.Powershell.ScriptAnalyzer.BuiltinRules
+{/// <summary>
+    /// UseShouldProcessCorrectly: Analyzes the ast to check that if the ShouldProcess attribute is present, the function calls ShouldProcess and vice versa.
+    /// </summary>
+    [Export(typeof(IScriptRule))]
+    public class UseShouldProcessForStateChangingFunctions : IScriptRule
+    {
+        /// <summary>
+        /// AnalyzeScript: Analyzes the ast to check if ShouldProcess is included in Advanced functions if the Verb of the function could change system state.
+        /// </summary>
+        /// <param name="ast">The script's ast</param>
+        /// <param name="fileName">The script's file name</param>
+        /// <returns>A List of diagnostic results of this rule</returns>
+        public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null) throw new ArgumentNullException(Strings.NullAstErrorMessage);
+
+            IEnumerable<Ast> funcDefAsts = ast.FindAll(testAst => testAst is FunctionDefinitionAst, true);
+            string supportsShouldProcess = "SupportsShouldProcess";
+            string trueString = "$true";
+            foreach (FunctionDefinitionAst funcDefAst in funcDefAsts)
+            {
+                string funcName = funcDefAst.Name;
+                bool hasShouldProcessAttribute = false;
+
+                if (funcName.StartsWith("Get-", StringComparison.OrdinalIgnoreCase) ||
+                    funcName.StartsWith("Stop-", StringComparison.OrdinalIgnoreCase)||
+                    funcName.StartsWith("New-", StringComparison.OrdinalIgnoreCase) ||
+                    funcName.StartsWith("Set-", StringComparison.OrdinalIgnoreCase) ||
+                    funcName.StartsWith("Update-", StringComparison.OrdinalIgnoreCase) ||
+                    funcName.StartsWith("Reset-", StringComparison.OrdinalIgnoreCase))
+                {
+                    IEnumerable<Ast> attributeAsts = funcDefAst.FindAll(testAst => testAst is NamedAttributeArgumentAst, true);
+                    if (funcDefAst.Body != null && funcDefAst.Body.ParamBlock != null
+                        && funcDefAst.Body.ParamBlock.Attributes != null &&
+                        funcDefAst.Body.ParamBlock.Parameters != null)
+                    {
+                        if (!funcDefAst.Body.ParamBlock.Attributes.Any(attr => attr.TypeName.GetReflectionType() == typeof (CmdletBindingAttribute)))
+                        {
+                            continue;
+                        }
+
+                        foreach (NamedAttributeArgumentAst attributeAst in attributeAsts)
+                        {
+                            if (attributeAst.ArgumentName.Equals(supportsShouldProcess, StringComparison.OrdinalIgnoreCase) && attributeAst.Argument.Extent.Text.Equals(trueString, StringComparison.OrdinalIgnoreCase))
```

Comment:
good catch - Quoc.

Yuting - pls look at the DSC Rule for mandatory parameters to check for attributes with/without $true